### ISSUE TITLE
mmod_loss consistently negative after a few steps with simple dataset

### ISSUE THREAD

**Issue by AndreaPisoni** [davisking/dlib#398]
As the title says, the loss gets negative quickly and the resulting network produces random results with a ton of false positive if I let the training run and let the loss go deep in the negative. 

I checked the output of 

```loss += truth->size()*options.loss_per_missed_target;
                for (auto&& x : *truth)
                {
                    if (!x.ignore)
                    {
                        point p = image_rect_to_feat_coord(input_tensor, x, sub);
                        loss -= out_data[p.y()*output_tensor.nc() + p.x()];

                        // compute gradient
                        g[p.y()*output_tensor.nc() + p.x()] = -scale;
                    }
                    else
                    {
                        // This box was ignored so shouldn't have been counted in the loss.
                        loss -= 1;
                    }
                }
```

Checking the loss value at line 579 of loss.h I can see the loss already gets negative at this point after a few thousands training steps, before even starting computing the truth hits, which means `loss -= out_data[p.y()*output_tensor.nc() + p.x()];` is probably subtracting more than it should. What does this operation represent exactly? I don't fully understand this and it would very useful to debug.

I have attached my bounding boxes here. Nothing special, a lot of boxes close to each other (so maybe something gets double counted in the process?) but very minimal overlap. 

[training_resized.txt](https://github.com/davisking/dlib/files/705778/training_resized.txt)

**Comment by davisking** [comment]
If you post a complete program I can run that reproduces the problem I'll take a look.

**Comment by davisking** [comment]
It occurred to me that the problem is probably the following.  Inside the mmod_options is a overlaps_nms object that is responsible for deciding if two boxes overlap for the purposes of applying non-max suppression during detection.  When you initially create the mmod_options object it will set the overlaps_nms to the strictest notion of overlap that doesn't reject any of the boxes in your training data.  So for example, if none of your training boxes overlap then the overlaps_nms will be set to consider any boxes that overlap at all, even by a single pixel, as overlapping and will not allow any overlapping boxes in the output after non-max suppression.   Or alternatively, if the most overlap present in your training data is a 10% overlap then it won't allow any more than 10% overlap in the output.  You get the idea.

So what could happen in your case is the overlaps_nms object is configured to hardly allow any overlap, maybe 1% in your case.  However, inside the training code there are two coordinate systems.  The coordinates in the input images and the coordinates in the feature maps output by the CNN.  The code needs to map between these coordinates and back and this image->cnn->image mapping is in general not perfect since multiple image locations typically get mapped to one cnn location (since cnn feaures are often spatially downsampled at least a little relative to the input image, which is in fact the case for the example cnn architecture shown in the example program).  This means that some of your images might contain impossible labelings given the setting of the overlaps_nms object.  In your case, you have a bunch of rectangles that are right next to each other but not really overlapping.  It could easily be the case that after the image->cnn->image mapping some of them do overlap and then get thrown out during non-max-suppression.  

I'm mentioning all this because the only way the MMOD loss function can go negative is essentially when the truth image labeling is impossible for the detector to output.  I have code in there that tries to validate the training data and give you an error if your boxes are "impossible" in this sense.  However, maybe the test isn't quite right and it's erroneously letting your data go by.  Although, assuming this is all true, the real problem is that the overlaps_nms is set too tight by default.

So the solution and test to see if this is happening is to change the mmod_options::overlaps_nms object to something less strict.  So set it to an overlap threshold of 0.5 for instance.  Does doing that make the problem go away?

**Comment by AndreaPisoni** [comment]
Good idea, I noticed that reducing the size of the boxes by 25% the issue disappeared so this might work. Let me try and get back with some results.

**Comment by AndreaPisoni** [comment]
Yeah I can confirm the error does not show up if I allow for 40% overlap

**Comment by davisking** [comment]
Sweet.  I'm going to make this a little less strict so hopefully others don't have the same problem.  In particular, if you instead of setting it to 0.4 replace the mmod_options::overlaps_nms setting with this:

```
double relax_amount = 0.95;
auto iou_thresh = options.overlaps_nms.get_iou_thresh() * relax_amount;
auto percent_covered_thresh = options.overlaps_nms.get_percent_covered_thresh() * relax_amount;
options.overlaps_nms = test_box_overlap(iou_thresh, percent_covered_thresh);
```

Does that work?  If so I'll make that the default setting mmod_options::overlaps_nms.

**Comment by AndreaPisoni** [comment]
I will try this, I am just wondering, should the relax amount be <1? I am thinking it was failing with ~0 and it works alright with 0.4, so the relax amount should probably allow for an higher threshold rather than a smaller one?

**Comment by davisking** [comment]
Oops, right, I forgot which way it interprets it.  Yes, it definitely needs to get bigger :)

So yeah, try a relax_amount of 1.05 or 1.1 or something like that.  You will also need to make sure the `_thresh` parameters don't get bigger than 1 because there are asserts about that.  Wrapping them in a `std::min(1.0,var)` statement should do the trick.

**Comment by AndreaPisoni** [comment]
Problem is if my boxes do not have any overlap at all  the noms iou threshold will stay at 0... should there be a baseline minimum overlap allowed?

**Comment by davisking** [comment]
Ah, I thought they overlapped a little.  But yeah, having a small baseline overlap is probably a good idea too.

**Comment by davisking** [comment]
I just pushed what we talked about to github, so this should be fixed by default now.
